src/crafting.cpp

```diff
@@ -59,40 +58,57 @@ recipe::recipe() :
 {
 }
 
-const recipe *find_recipe( std::string id )
+void recipe_dictionary::add( recipe* rec )
 {
-    for( auto recipe_list : recipes ) {
-        for( auto recipe : recipe_list.second ) {
-            if( recipe->ident == id ) {
-                return recipe;
-            }
-        }
-    }
-    return nullptr;
+    recipes.push_back( rec );
+    add_to_component_lookup( rec );
+    by_name[rec->ident] = rec;
+    by_index[rec->id] = rec;
+    by_category[rec->cat].push_back( rec );
 }
 
-void add_to_component_lookup(recipe* r)
+void recipe_dictionary::remove( recipe* rec ) {
+    recipes.remove( rec );
+    remove_from_component_lookup( rec );
+    by_name.erase( rec->ident );
+    by_index.erase( rec->id );
+    by_category.erase( rec->cat );
+}
+
+void recipe_dictionary::add_to_component_lookup( recipe* r )
 {
     std::unordered_set<itype_id> counted;
-    for (const auto &comp_choices : r->requirements.components) {
-        for (const item_comp &comp : comp_choices) {
-            if (counted.count(comp.type)) {
+    for( const auto &comp_choices : r->requirements.components ) {
+        for( const item_comp &comp : comp_choices ) {
+            if( counted.count(comp.type) ) {
                 continue;
             }
-            counted.insert(comp.type);
-            recipes_by_component[comp.type].push_back(r);
+            counted.insert( comp.type );
+            by_component[comp.type].push_back( r );
         }
     }
 }
 
-void remove_from_component_lookup(recipe* r)
+void recipe_dictionary::remove_from_component_lookup( recipe* r )
 {
-    for (auto &map_item : recipes_by_component) {
+    for( auto &map_item : by_component ) {
         std::vector<recipe *> &rlist = map_item.second;
-        rlist.erase(std::remove(rlist.begin(), rlist.end(), r), rlist.end());
+        rlist.erase( std::remove( rlist.begin(), rlist.end(), r ), rlist.end() );
     }
 }
 
+void recipe_dictionary::clear()
+{
+    by_component.clear();
+    by_name.clear();
+    by_index.clear();
+    by_category.clear();
+    for( auto &recipe : recipes ) {
+        delete recipe;
```

Comment:
I'm used to java, which doesn't have deconstructors, so I wanted to use them (loving syntactic sugar). In a way it makes sense to have the recipe remove itself. But it does feel very hacky. I'll change it.